src/Raven.Server/Documents/Handlers/Processors/Indexes/AbstractIndexHandlerProcessorForTerms.cs

```diff
@@ -0,0 +1,78 @@
+﻿using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Raven.Client.Documents.Operations.Indexes;
+using Raven.Client.Http;
+using Raven.Server.Documents.Queries;
+using Raven.Server.Json;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Web;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Handlers.Processors.Indexes
+{
+    internal abstract class AbstractIndexHandlerProcessorForTerms<TRequestHandler, TOperationContext> : AbstractHandlerProcessor<TRequestHandler, TOperationContext>
+        where TRequestHandler : RequestHandler
+        where TOperationContext : JsonOperationContext
+    {
+        protected AbstractIndexHandlerProcessorForTerms([NotNull] TRequestHandler requestHandler, [NotNull] JsonContextPoolBase<TOperationContext> contextPool)
+            : base(requestHandler, contextPool)
+        {
+        }
+
+        protected abstract ValueTask<TermsQueryResultServerSide> GetTermsAsync(string indexName, string field, string fromValue, int pageSize, long? resultEtag);
+
+        protected abstract long? GetLongFromHeaders(string name);
+
+        public override async ValueTask ExecuteAsync()
+        {
+            using (RequestHandler.ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var field = RequestHandler.GetQueryStringValueAndAssertIfSingleAndNotEmpty("field");
+                var indexName = RequestHandler.GetQueryStringValueAndAssertIfSingleAndNotEmpty("name");
+                var fromValue = RequestHandler.GetStringQueryString("fromValue", required: false) ?? "";
+                var pageSize = RequestHandler.GetIntValueQueryString("pageSize", required: false) ?? int.MaxValue;
+                var resultEtag = GetLongFromHeaders("If-None-Match");
```

Comment:
We need `resultEtag` for `ShardedGetTermsOperation` which now inherits from `IShardedReadOperation`. Then I'm doing the above instead of sending it from both (sharded & non-sharded constructors)